pkg/kwok/controllers/metrics/metrics.go

```diff
@@ -0,0 +1,136 @@
+limitations under the License.
+*/
+
+// Package metrics provides node handler used by controllers to publish metrics
+package metrics
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"sigs.k8s.io/kwok/pkg/apis/internalversion"
+	kwokmetrics "sigs.k8s.io/kwok/pkg/kwok/metrics"
+	"sigs.k8s.io/kwok/pkg/kwok/metrics/cel"
+	"sigs.k8s.io/kwok/pkg/utils/maps"
+)
+
+// NodeHandler handles updating metrics for each node that controller reconcile on
+type NodeHandler struct {
+	// state
+	nodeContainersCounter maps.SyncMap[string, float64]
+	nodeCache             maps.SyncMap[string, *corev1.Node]
+
+	// configuration
+	Metrics []*internalversion.Metric
+}
+
+// UpdateCache updates node cache
+func (h *NodeHandler) UpdateCache(node *corev1.Node) {
+	h.nodeCache.Store(node.Name, node)
+}
+
+// AddContainersCount adds the number of containers started on a node
+func (h *NodeHandler) AddContainersCount(nodeName string, count int) {
+	actual, loaded := h.nodeContainersCounter.LoadOrStore(nodeName, float64(count))
+	if !loaded {
+		return
+	}
+	h.nodeContainersCounter.Store(nodeName, actual+float64(count))
+}
+
+// ContainersCount returns the number of containers started on a node
+func (h *NodeHandler) ContainersCount(nodeName string) float64 {
+	v, _ := h.nodeContainersCounter.Load(nodeName)
+	return v
+}
+
+// Publish publishes node metrics to all custom metrics registries
+func (h *NodeHandler) Publish(nodeName string) error {
+	if len(h.Metrics) == 0 {
+		return nil
+	}
+	node, ok := h.nodeCache.Load(nodeName)
+	if !ok {
+		return fmt.Errorf("node %q not found in cache", nodeName)
+	}
+
+	var errs []error
+	for _, metric := range h.Metrics {
+		eval, err := cel.NewNodeEvaluator(node, h.ContainersCount)
+		if err != nil {
+			return fmt.Errorf("failed to instantiate evaluator for a node %q: %w", node.Name, err)
+		}
+
+		reg := kwokmetrics.CustomMetric(metric.Name)
+		for i := range metric.Spec.Metrics {
+			m := &metric.Spec.Metrics[i]
+			switch m.Kind {
+			case kwokmetrics.KindGauge:
+				g, err := reg.Gauge(m, node)
+				if err != nil {
+					errs = append(errs, fmt.Errorf("failed to get/create gauge for metric %q with configuration name %q: %w", metric.Name, m.Name, err))
+					continue
+				}
+				val, err := eval.Evaluate(m.Value)
+				if err != nil {
+					errs = append(errs, fmt.Errorf("failed to evaluate expression %q in metric %q: %w", m.Value, metric.Name, err))
+				}
+				g.Set(val)
+			case kwokmetrics.KindCounter:
+				c, err := reg.Counter(m, node)
+				if err != nil {
+					errs = append(errs, fmt.Errorf("failed to get/create counter for metric %q with configuration name %q: %w", metric.Name, m.Name, err))
+					continue
+				}
+				val, err := eval.Evaluate(m.Value)
+				if err != nil {
+					errs = append(errs, fmt.Errorf("failed to evaluate expression %q in metric %q: %w", m.Value, metric.Name, err))
+					continue
+				}
+				c.Add(val)
```

Comment:
https://github.com/kubernetes-sigs/kwok/pull/555#discussion_r1202237446

I've found it a bit tricky, as it causes each update to increase the current total number of containers.

Reply:
Oh but it should be 2 right? Because kubelet started 2 containers. One once you created a pod, and another one once you modified it, it re-created a container. My understanding of this metric is to say how many containers did the kubelet start, not how many of them are running at this time

Reply:
Your understanding is correct.

But the `val` value is the cumulative total of containers started.

Each execution of `c.Add(val)` will cause the value in the indicator to increase the cumulative number of containers started.

Reply:
I have been thinking how can we implement this with the ContainerFunc and I don't know how effectively to solve it.
I had another idea that is possibly not that good (I don't know nearly this as much as you do). Can we leverage stages? 

Since the pod controller keeps track of the resource version and it computes the stage transition, can we leverage that? When the pod controller is about to set the stage to `ContainersReady`, we increment the count. That way, we can avoid allocating memory and trying to figure out if this container is about to be executed (there is already logic there that keeps track of the stage transitions). And also, we don't need to react on events. As soon as the controller knows it should transition the pod to `ContainersReady`, that is when we increment the count. The `Add` event can be then avoided (cache re-sync), and the metric would be correct (what if the stage applied is to simulate a faulty container image that couldn't be pulled from the container registry, for example. Then the number of container started should not be incremented).
What do you think?

Reply:
🤣I think it should work. I will give it a try.

Reply:
Found a problem
THe value: `( unixSecond(now()) - unixSecond(node.metadata.creationTimestamp) ) / 10.0` can't work since the creationTimestamp is a string, and unixSeconds accepts Time. In order to be able to work with this interface, should we change now() to return a string? Then unixSecond will accept that, parse it as time and return seconds.
We can also introduce another function called time for example, to parse string to a time

What would you prefer?

Reply:
https://github.com/kubernetes-sigs/kwok/commit/3cb4ae98ac14279f52351e3f914ea321bb91c095

I've refactored, which should solve both of your problems. 😄
But I've found a new problem with the Histogram implementation, it doesn't work as expected, and I think the configuration needs to be changed.

Reply:
Did I write over it with rebase :disappointed: ? I'll see the diff and try to apply it back, sorry :smile:

Reply:
I see some comments, are you refactoring this now? It would be awesome if you push it to this branch :relaxed:

Reply:
done 😄